lib/dry/schema/macros/key.rb

```diff
@@ -21,6 +48,15 @@ def to_rule
         def to_ast
           [:predicate, [:key?, [[:name, name], [:input, Undefined]]]]
         end
+
+        private
+
+        def extract_type_spec(*args)
+          type_spec = args[0].is_a?(Symbol) && !args[0].to_s.end_with?(QUESTION_MARK) && args[0]
```

Comment:
OK, so I like what we're doing here by allowing the first arg of `value`/`filled`/`maybe` to act as the type, but it seems we have some creeping inconsistencies here?

When we use the _explicit_ `type` API, e.g.

```ruby
required(:age).type(:int).filled(gt?: 18)
```

We allow _any_ value to be passed to `#type` and then it's just directly passed to `schema_dsl.set_type` from there.

So this behaviour on `type` would also allow something like this, I'm assuming:

```ruby
required(:age).type(Types::MyCustomIntegerType).filled(gt?: 18)
```

Am I right with that?

So then we come to this alternative approach to specifying types, via `#extract_type_spec, where we infer a type specification from the first arg passed to `#value`, we have a much stricter set of rules, i.e. it _must_ be a symbol not ending in a question mark.

This would obviously prevent e.g. `MyCustomIntegerType` from being used as the first arg here. And I presume it'd also prevent type specs being passed like `[:nil, MyCustomIntegerType]` where we want a value to be nullable.

Would it make sense here to tweak the logic for this arg check a little, perhaps to something like this?

```ruby
type_spec = args[0]
type_spec = nil if type_spec.is_a?(Symbol) && type_spec.to_s.end_with?(QUESTION_MARK)
```

I'm sure we can find a more succinct way to do this in code, but I think this naive first pass actually expresses the _logic_ pretty clearly, i.e.

1. Let's presume the first arg is going to be a type spec (this is what'll let us permit _any_ kind of object acting as the spec)
2. But if it turns out to be a symbol ending with a question mark, let's take it back, that's not a type spec, that's actually just an argument-less predicate

So this would permit both `value(:int, gt?: 18)` _and_ `value(Types::MyCustomInteger, gt?: 18)`, as well as any other type spec format.

Does... that make sense?

But I also wonder, does permitting the first arg to `#value` to be a type spec, especially if we loosen the arg check like I suggest, actually render the entire `#type` method redundant? 

I guess `#type` is there for (a) when we're not actually providing any `#value` rules for a key, and (b) making things more explicit/clearer by letting us separate the type spec definition rather than mixing it with the predicates (i.e. more a stylistic thing more than anything else)

Am I overlooking anything in the nuances of handling types here?

Reply:
FWIW, I read back over our conversation about this in Gitter, and it seems like we came to the conclusion that _this_ was a good DSL:

```ruby
required(:age).filter(format?: /\d+/).value(:integer, gt?: 18)
```

And that it would effectively be a _synonym_ for:

```ruby
required(:age).filter(format?: /\d+/).type(:integer).value(gt?: 18)
```

So I think this is good motivation for us to make our type handling consistent across both variations of that expression, i.e. allow the same type spec arg in both.

Reply:
@timriley this was a major oversight on my behalf! thanks for spotting this, it's fixed now